backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/utils/DedicatedCpuPinningHelper.java

```diff
@@ -0,0 +1,152 @@
+
+import org.ovirt.engine.core.common.businessentities.CpuPinningPolicy;
+import org.ovirt.engine.core.common.businessentities.VDS;
+import org.ovirt.engine.core.common.businessentities.VM;
+import org.ovirt.engine.core.common.businessentities.VdsCpuUnit;
+import org.ovirt.engine.core.compat.Guid;
+
+public class DedicatedCpuPinningHelper {
+
+    public static synchronized boolean isDedicatedCpuPinningPossibleAtHost(List<VdsCpuUnit> cpuTopology,
+            Map<Guid, List<VdsCpuUnit>> vmToPendingPinnings,
+            VM vm, VDS host) {
+
+        if (vm.getCpuPinningPolicy() != CpuPinningPolicy.DEDICATED) {
+            // TODO: Implementation for siblings and isolate threads
+            return false;
+        }
+
+        pinPendingExclusiveCpus(cpuTopology, vmToPendingPinnings);
+
+        int cpus = vm.getNumOfCpus();
+        int cpusLeft = cpus;
+        int hostCoresPerSocket = host.getCpuCores() / host.getCpuSockets();
+        int hostThreadsPerCore = host.getCpuThreads() / host.getCpuCores();
+
+        for (int socket = 0; socket < host.getCpuSockets(); socket++) {
+            if (hostCoresPerSocket * hostThreadsPerCore >= cpus && getFreeCpusInSocket(cpuTopology, socket).size() < cpus) {
+                // we should fit into one socket
+                continue;
+            }
+            for (int core = 0; core < hostCoresPerSocket; core++) {
+                int sharedCpusInCore = getFreeCpusInCore(cpuTopology, socket, core).size();
+                if (cpusLeft <= 0) {
+                    unPinPendingExclusiveCpus(cpuTopology, vmToPendingPinnings);
+                    return true;
+                }
+                cpusLeft-=sharedCpusInCore;
+            }
+        }
+        unPinPendingExclusiveCpus(cpuTopology, vmToPendingPinnings);
+        return cpusLeft == 0;
+    }
+
+    // TODO use a new lock or synchornized on VdsManager instance?
+    private static void pinPendingExclusiveCpus(List<VdsCpuUnit> cpuTopology, Map<Guid, List<VdsCpuUnit>> vmToPendingPinning) {
+        for (var vmToPendingPinningEntry : vmToPendingPinning.entrySet()) {
+            vmToPendingPinningEntry.getValue().forEach(vdsCpuUnit -> vdsCpuUnit.setVmId(vmToPendingPinningEntry.getKey()));
+        }
+        //cpuTopology.stream().filter(cpu -> vmToPendingPinning.containsValue(cpu)).forEach(cpu -> cpu.setPinned(true));
+    }
+
+    private static void unPinPendingExclusiveCpus(List<VdsCpuUnit> cpuTopology, Map<Guid, List<VdsCpuUnit>> vmToPendingPinning) {
+        for (var vmToPendingPinningEntry : vmToPendingPinning.entrySet()) {
+            vmToPendingPinningEntry.getValue().forEach(vdsCpuUnit -> vdsCpuUnit.setVmId(null));
+        }
+        //cpuTopology.stream().filter(cpu -> vmToPendingPinning.containsValue(cpu)).forEach(cpu -> cpu.setPinned(false));
+    }
+
+    public static synchronized List<VdsCpuUnit> allocateDedicatedCpus(List<VdsCpuUnit> cpuTopology, VM vm, Map<Guid, List<VdsCpuUnit>> vmToPendingPinnings, VDS host) {
```

Comment:
The same as above - please add a comment, it is hard to follow just from the code.